File: src/test/java/com/marcelomachado/urlshortener/controller/UrlShortenerControllerTest.java
package com.marcelomachado.urlshortener.controller;

import com.marcelomachado.urlshortener.service.UrlShortenerService;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.ResponseEntity;

@SpringBootTest
class urlShortenerControllerTests {

    private int BAD_REQUEST = 400;
    private int CREATED = 201;

    @Mock
    private UrlShortenerService urlShortenerService;

    @Autowired
    private UrlShortenerController urlShortenerController;

    @BeforeEach
    void setup() {
        urlShortenerController = new UrlShortenerController(urlShortenerService);
    }

    @Test
    @DisplayName("When receive a valid url should return a response with CREATED status")
    void postUrl_shouldReturnUrl() {
        String url = "https://www.marcelomachado.com";
        Mockito.when(urlShortenerService.shortenUrl(Mockito.any())).thenReturn("index/hash");

        ResponseEntity<String> response = urlShortenerController.postUrl(url);

        Assertions.assertEquals(CREATED, response.getStatusCode().value());
        Assertions.assertEquals("index/hash", response.getBody());
    }

    @Test
    @DisplayName("When receive an incomplete url should return a response with BADREQUEST status")
    void postUrl_shouldReturnBadRequest_incompleteURL() {
        String url = "marcelomachado.com";

        ResponseEntity<String> response = urlShortenerController.postUrl(url);

        Assertions.assertEquals(BAD_REQUEST, response.getStatusCode().value());
    }

    @Test
    @DisplayName("When receive an invalid url should return a response with BADREQUEST status")
    void postUrl_shouldReturnBadRequest_invalidURL() {
        String url = "marcelomachado";

        ResponseEntity<String> response = urlShortenerController.postUrl(url);

        Assertions.assertEquals(BAD_REQUEST, response.getStatusCode().value());
    }

}

File: src/test/kotlin/com/marcelomachado/urlshortener/controller/UrlShortenerControllerTest.kt
package com.marcelomachado.urlshortener.controller

import com.marcelomachado.urlshortener.service.UrlShortenerService
import io.mockk.clearMocks
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import reactor.core.publisher.Mono
import kotlin.test.assertEquals

@SpringBootTest
class UrlShortenerControllerTest {

    @Autowired
    private lateinit var urlShortenerController: UrlShortenerController

    private lateinit var urlShortenerService: UrlShortenerService

    private val validUrl: String = "https://marcelomachado.com"
    private val invalidUrl: String = "invalid-url"
    private val shortenMockUrl: String = "url-shortened"
    private val badRequestStatusCode: Int = 400
    private val createdStatusCode: Int = 201

    @BeforeEach
    fun setup() {
        urlShortenerService = mockk()

        urlShortenerController = UrlShortenerController(urlShortenerService)

        every { urlShortenerService.shortenUrl(validUrl) } returns Mono.just(shortenMockUrl)
    }

    @AfterEach
    fun tearDown() {
        clearMocks(urlShortenerService)
    }

    @Test
    @DisplayName("Should return 400 when invalid url")
    fun postUrl_ShouldReturnBadRequest_WhenInvalidUrl() {
        val response = urlShortenerController.postUrl(invalidUrl)

        assertEquals(badRequestStatusCode, response.block()?.statusCode?.value())
    }

    @Test
    @DisplayName("Should return 201 when valid url")
    fun postUrl_ShouldReturnOk_WhenValidUrl() {
        val response = urlShortenerController.postUrl(validUrl).block()

        assertEquals(createdStatusCode, response?.statusCode?.value())
        assertEquals(shortenMockUrl, response?.body)

        verify(exactly = 1) { urlShortenerService.shortenUrl(validUrl) }
    }

}
